From 4f09e785c0a3d0ce050e9eff8f28af0c82e27620 Mon Sep 17 00:00:00 2001
From: gtsoul-tech <56584633+gtsoul-tech@users.noreply.github.com>
Date: Wed, 13 Nov 2024 10:43:23 +0200
Subject: [PATCH] Fix regression error #317 and add unit test (#318)

Revert the code that produced the regression error in #317
Add the regression error to a unit test regressions.cpp along with the rebar tests

https://github.com/VectorCamp/vectorscan/commit/4f09e785c0a3d0ce050e9eff8f28af0c82e27620

---------

Co-authored-by: gtsoul-tech <gtsoulkanakis@gmail.com>
---
 src/nfa/shufti_simd.hpp                       | 12 +----
 unit/CMakeLists.txt                           |  2 +-
 .../{rebar_tests.cpp => regressions.cpp}      | 50 ++++++++++++++++++-
 3 files changed, 52 insertions(+), 12 deletions(-)
 rename unit/hyperscan/{rebar_tests.cpp => regressions.cpp} (83%)

diff --git a/src/nfa/shufti_simd.hpp b/src/nfa/shufti_simd.hpp
index bdb0ff9fe..1a00b87b9 100644
--- a/src/nfa/shufti_simd.hpp
+++ b/src/nfa/shufti_simd.hpp
@@ -244,16 +244,8 @@ const u8 *shuftiDoubleExecReal(m128 mask1_lo, m128 mask1_hi, m128 mask2_lo, m128
     // finish off tail
 
     if (d != buf_end) {
-        SuperVector<S> chars = SuperVector<S>::Zeroes();
-        const u8 *end_buf;
-        if (buf_end - buf < S) {
-          memcpy(&chars.u, buf, buf_end - buf);
-          end_buf = buf;
-        } else {
-          chars = SuperVector<S>::loadu(buf_end - S);
-          end_buf = buf_end - S;
-        }
-        rv = fwdBlockDouble(wide_mask1_lo, wide_mask1_hi, wide_mask2_lo, wide_mask2_hi, chars, end_buf);
+        SuperVector<S> chars = SuperVector<S>::loadu(d);
+        rv = fwdBlockDouble(wide_mask1_lo, wide_mask1_hi, wide_mask2_lo, wide_mask2_hi, chars, d);
         DEBUG_PRINTF("rv %p \n", rv);
         if (rv && rv < buf_end) return rv;
     }